### PR TITLE
Split posts and comments into separate Elasticsearch indices

### LINKED ISSUE #1337: Split posts and comments into separate Elasticsearch indices
- Posts, comments should each get their own index and mapping
- Refactor index creation, aliasing, & updates

### PR DESCRIPTION
**Pull request by mbertrand**
#### Pre-Flight checklist
- [x] Testing
  - [x] Code is tested
  - [x] Changes have been manually tested

#### What are the relevant tickets?
Closes #1337 

#### What's this PR do?
- Creates (and recreates) separate indices for posts and comments
- Creates an alias to allow searching both indices simultaneously
- Defines separate mappings for posts and comments
- Adds `channel_name` to both post and comment mappings

#### How should this be manually tested?
- Run `python manage.py recreate_index`
- After it's completed, posts should be found in a `discussions_local_post_default` index, comments in a `discussions_local_comment_default` index, and there should be an alias `discussions_local_all_default`.  The original `discussions_local_default` index from master branch will still exist unless you manually delete it or never created it in the first place.
- If you post the following to `http://localhost:9101/discussions_local_all_default/_search/`, you should get a list of posts back:
  ```javascript
  {
    "query": {
      "term": {
        "object_type": "post"
      }
    }
  }
  ```
- If you post the above but replace `post` with `comment`, you should get a list of comments back.

- Create a new post, then add some comments.  The new post should be added to the post index only, the comments should be added to the comment index only.  
- 'Remove' a post that has comments, then search for the records in Elasticsearch.  The comments should have `parent_post_removed: true`

### CODE REVIEW

File: search/indexing_api.py [hunk 1]
```diff
@@ -44,25 +43,35 @@
...
     """
     Wipe and recreate index and mapping. No indexing is done.
 
     Args:
         index_name (str): The name of the index to clear
         skip_mapping (bool): If true, don't set any mapping
+        object_type(str): The type of document (post, comment)
     """
+    if not object_type or object_type not in VALID_OBJECT_TYPES:
+        raise TypeError('A valid object type must be specified when clearing and creating an index')
```
Comment (noisecapella) [inline search/indexing_api.py]: This is pretty nitpicky but we should use `ValueError` here. According to [the docs](https://docs.python.org/3/library/exceptions.html#TypeError) `TypeError` is only if the Python types don't match with each other

File: search/indexing_api.py [hunk 2]
```diff
@@ -263,8 +292,22 @@ def index_post_with_comments(post_id):
...
+        _, errors = bulk(
+            conn,
+            sync_comments(serialize_bulk_comments(post)),
+            index=alias,
+            doc_type=GLOBAL_DOC_TYPE,
+            # Adjust chunk size from 500 depending on environment variable
+            chunk_size=settings.ELASTICSEARCH_INDEXING_CHUNK_SIZE,
+        )
+        if len(errors) > 0:
+            raise ReindexException("Error during bulk insert: {errors}".format(
```
Comment (noisecapella) [inline search/indexing_api.py]: Can you change the message here and above to say whether it was comment or post we're indexing?

File: search/indexing_api.py [hunk 3]
```diff
@@ -263,8 +292,22 @@ def index_post_with_comments(post_id):
...
+                errors=errors
+            ))
 
-def create_backing_index():
+
+def create_backing_index(object_type):
     """
     Start the reindexing process by creating a new backing index and pointing the reindex alias toward it
```
Comment (noisecapella) [inline search/indexing_api.py]: Can you add `object_type` to the docstring?

File: search/indexing_api.py [hunk 4]
```diff
@@ -291,7 +334,7 @@ def create_backing_index():
     return new_backing_index
 
 
-def switch_indices(backing_index):
+def switch_indices(backing_index, object_type):
     """
     Switch the default index to point to the backing index, and delete the reindex alias
```
Comment (noisecapella) [inline search/indexing_api.py]: Can you add `object_type` to the docstring?

File: search/serializers_test.py [hunk 5]
```diff
@@ -133,8 +132,7 @@ def test_serialize_bulk_post_and_comments(mocker, patched_base_post_serializer,
         replies=[inner_comment_mock]
     )
     post_mock = mocker.MagicMock(comments=[outer_comment_mock])
-    assert list(serialize_bulk_post_and_comments(post_mock)) == [
-        serialize_post_for_bulk(post_mock),
+    assert list(serialize_bulk_comments(post_mock)) == [
```
Comment (noisecapella) [inline search/serializers_test.py]: Can you change the name of this test? Do we have a test for serializing posts?

File: search/tasks.py [hunk 6]
```diff
@@ -159,13 +158,13 @@ def start_recreate_index(self):
...
             index_channels,
-            finish_recreate_index.s(new_backing_index),
+            finish_recreate_index.s(new_backing_indices),
         )
     )
 
 
 @app.task
-def finish_recreate_index(results, backing_index):
+def finish_recreate_index(results, backing_indices):
```
Comment (noisecapella) [inline search/tasks.py]: Can you update the docstring?

File: search/connection.py [hunk 7]
```diff
@@ -52,56 +53,69 @@ def get_conn(*, verify=True):
...
 
     Returns:
         list of str: Aliases which exist
     """
+    if not object_types:
+        object_types = VALID_OBJECT_TYPES
     conn = get_conn(verify=False)
     return [
-        alias for alias in [get_default_alias_name(), get_reindexing_alias_name()]
+        alias for x in [
```
[the remark this answers is not in the document]
Comment (mbertrand) [reply search/connection.py]: ✅

File: search/indexing_api.py [hunk 8]
```diff
@@ -44,25 +43,35 @@
...
     'post_title': {'type': 'text'},
-    'post_link_url': {'type': 'keyword'},
-    'post_link_thumbnail': {'type': 'keyword'},
-    'num_comments': {'type': 'long'},
-    'comment_id': {'type': 'keyword'},
-    'parent_comment_id': {'type': 'keyword'},
+}
+
+MAPPING = {
+    POST_TYPE: dict(BASE_CONTENT_TYPE, **{
```
[the remark this answers is not in the document]
Comment (mbertrand) [reply search/indexing_api.py]: ✅

File: search/indexing_api.py [hunk 9]
```diff
@@ -44,25 +43,35 @@
...
+def clear_and_create_index(*, index_name=None, skip_mapping=False, object_type=None):
     """
     Wipe and recreate index and mapping. No indexing is done.
 
     Args:
         index_name (str): The name of the index to clear
         skip_mapping (bool): If true, don't set any mapping
+        object_type(str): The type of document (post, comment)
     """
+    if not object_type or object_type not in VALID_OBJECT_TYPES:
```
[the remark this answers is not in the document]
Comment (mbertrand) [reply search/indexing_api.py]: ✅

File: search/indexing_api.py (hunk 1 above)
Comment (mbertrand) [reply search/indexing_api.py]: ✅

File: search/indexing_api.py (hunk 2 above)
Comment (mbertrand) [reply search/indexing_api.py]: ✅

File: search/indexing_api.py (hunk 3 above)
Comment (mbertrand) [reply search/indexing_api.py]: ✅

File: search/indexing_api.py (hunk 4 above)
Comment (mbertrand) [reply search/indexing_api.py]: ✅

File: search/serializers_test.py (hunk 5 above)
Comment (mbertrand) [reply search/serializers_test.py]: ✅ , `test_serialize_post_for_bulk`

File: search/tasks.py (hunk 6 above)
Comment (mbertrand) [reply search/tasks.py]: ✅

**Comment by mbertrand** [comment]
@noisecapella I added `channel_name` to the mappings, and fixed a bug I noticed where `channel_title` was getting mixed up with `channel_name` in the comments index.

**Comment by noisecapella** [comment]
Thanks, functionality is looking great. Looking over code now

**Pull request merged**